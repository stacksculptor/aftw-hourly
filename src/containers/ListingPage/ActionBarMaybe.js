import React from 'react';
import { bool, oneOfType, object } from 'prop-types';
import { FormattedMessage } from 'react-intl';
import classNames from 'classnames';
import { LISTING_STATE_PENDING_APPROVAL, LISTING_STATE_CLOSED, propTypes } from '../../util/types';
import { NamedLink } from '../../components';
import EditIcon from './EditIcon';

import css from './ListingPage.css';

export const ActionBarMaybe = props => {
  const { isOwnListing, listing, editParams } = props;
  const state = listing.attributes.state;
  const isPendingApproval = state === LISTING_STATE_PENDING_APPROVAL;
  const isClosed = state === LISTING_STATE_CLOSED;

  if (isOwnListing) {
    let ownListingTextTranslationId = 'ListingPage.ownListing';

    if (isPendingApproval) {
      ownListingTextTranslationId = 'ListingPage.ownListingPendingApproval';
    } else if (isClosed) {
      ownListingTextTranslationId = 'ListingPage.ownClosedListing';
    }

    const ownListingTextClasses = classNames(css.ownListingText, {
      [css.ownListingTextPendingApproval]: isPendingApproval,
    });

    return (
      <div className={css.actionBar}>
        <p className={ownListingTextClasses}>
          <FormattedMessage id={ownListingTextTranslationId} />
        </p>
        <NamedLink className={css.editListingLink} name="EditListingPage" params={editParams}>
          <EditIcon className={css.editIcon} />
          <FormattedMessage id="ListingPage.editListing" />
        </NamedLink>
      </div>
    );
  } else if (isClosed) {
    return (
      <div className={css.actionBar}>
        <p className={css.closedListingText}>
          <FormattedMessage id="ListingPage.closedListing" />
        </p>
      </div>
    );
  }
  return null;
};

ActionBarMaybe.propTypes = {
  isOwnListing: bool.isRequired,
  listing: oneOfType([propTypes.listing, propTypes.ownListing]).isRequired,
  editParams: object.isRequired,
};

ActionBarMaybe.displayName = 'ActionBarMaybe';

export default ActionBarMaybe;
