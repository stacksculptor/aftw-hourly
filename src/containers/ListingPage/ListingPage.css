.pageRoot {}

.listing {
  display: block;
}

.listing {
  padding-bottom: 80px;
}

.header {
  display: flex;
  flex-direction: row;
  padding: 1rem;
}

.title,
.price {
  font-size: 2em;
  margin: 0.67em 0;
}

.title {
  flex-grow: 1;
}

.mainImage {
  width: 100%;
}

.thumbnailContainer {
  display: flex;
  flex-direction: row;
  padding: 2rem 1.5rem;
}

.squareWrapper {
  display: block;
  position: relative;
  width: 100px;
  margin: 0 0.5rem;
}

/* Firefox doesn't support image aspect ratio inside flexbox */
.aspectWrapper {
  padding-bottom: 75%; /* 4:3 Aspect Ratio */
}

.thumbnail {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

.description {
  margin: 2rem;
}

.map {
  margin-top: 3rem;
  height: 300px;
  width: 100%;
  background-color: #eee;
}

.bookingForm {
  margin: 0 1rem;
}

.openBookingForm {
  position: fixed;
  bottom: 0;
  left: 0;
  width: 100vw;
  height: 80px;
  padding: 1rem;
  background-color: white;
  border-top: solid 1px #ccc;

  /* Mobile frame fix */
  max-width: 375px;
  left: 50%;
  transform: translateX(-50%);
}

.bookingModalTitle {
  margin: 1rem 0;
}
