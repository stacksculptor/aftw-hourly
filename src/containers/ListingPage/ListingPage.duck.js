import { types } from '../../util/sdkLoader';
import { storableError } from '../../util/errors';
import { addMarketplaceEntities } from '../../ducks/marketplaceData.duck';
import { updatedEntities, denormalisedEntities } from '../../util/data';
import { fetchCurrentUser } from '../../ducks/user.duck';

// ================ Action types ================ //

export const SHOW_LISTING_REQUEST = 'app/ListingPage/SHOW_LISTING_REQUEST';
export const SHOW_LISTING_ERROR = 'app/ListingPage/SHOW_LISTING_ERROR';

export const FETCH_REVIEWS_REQUEST = 'app/ListingPage/FETCH_REVIEWS_REQUEST';
export const FETCH_REVIEWS_SUCCESS = 'app/ListingPage/FETCH_REVIEWS_SUCCESS';
export const FETCH_REVIEWS_ERROR = 'app/ListingPage/FETCH_REVIEWS_ERROR';

// ================ Reducer ================ //

const initialState = {
  id: null,
  showListingError: null,
  reviews: [],
  fetchReviewsError: null,
};

const listingPageReducer = (state = initialState, action = {}) => {
  const { type, payload } = action;
  switch (type) {
    case SHOW_LISTING_REQUEST:
      return { ...state, id: payload.id, showListingError: null };
    case SHOW_LISTING_ERROR:
      return { ...state, showListingError: payload };
    case FETCH_REVIEWS_REQUEST:
      return { ...state, fetchReviewsError: null };
    case FETCH_REVIEWS_SUCCESS:
      return { ...state, reviews: payload };
    case FETCH_REVIEWS_ERROR:
      return { ...state, fetchReviewsError: payload };
    default:
      return state;
  }
};

export default listingPageReducer;

// ================ Action creators ================ //

export const showListingRequest = id => ({
  type: SHOW_LISTING_REQUEST,
  payload: { id },
});

export const showListingError = e => ({
  type: SHOW_LISTING_ERROR,
  error: true,
  payload: e,
});

export const fetchReviewsRequest = () => ({ type: FETCH_REVIEWS_REQUEST });
export const fetchReviewsSuccess = reviews => ({ type: FETCH_REVIEWS_SUCCESS, payload: reviews });
export const fetchReviewsError = error => ({ type: FETCH_REVIEWS_ERROR, payload: error });

// ================ Thunks ================ //

export const showListing = listingId => (dispatch, getState, sdk) => {
  dispatch(showListingRequest(listingId));
  dispatch(fetchCurrentUser());
  return sdk.listings
    .show({ id: listingId, include: ['author', 'author.profileImage', 'images'] })
    .then(data => {
      dispatch(addMarketplaceEntities(data));
      return data;
    })
    .catch(e => {
      dispatch(showListingError(storableError(e)));
    });
};

export const fetchReviews = listingId => (dispatch, getState, sdk) => {
  return sdk.reviews
    .query({ listing_id: listingId, include: ['author', 'author.profileImage'] })
    .then(response => {
      const entities = updatedEntities({}, response.data);
      const reviewIds = response.data.data.map(d => d.id);
      const denormalized = denormalisedEntities(entities, 'review', reviewIds);
      dispatch(fetchReviewsSuccess(denormalized));
    })
    .catch(e => {
      dispatch(fetchReviewsError(e));
    });
};

export const loadData = params => dispatch => {
  const listingId = new types.UUID(params.id);

  return Promise.all([dispatch(showListing(listingId)), dispatch(fetchReviews(listingId))]);
};
