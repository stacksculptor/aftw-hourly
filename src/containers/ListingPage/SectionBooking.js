import React from 'react';
import { FormattedMessage } from 'react-intl';
import { ModalInMobile, Button } from '../../components';
import { BookingDatesForm } from '../../forms';

import css from './ListingPage.css';

// This defines when ModalInMobile shows content as Modal
const MODAL_BREAKPOINT = 1023;

const SectionBooking = props => {
  const {
    listing,
    isOwnListing,
    isClosed,
    isBook,
    unitType,
    price,
    formattedPrice,
    priceTitle,
    handleBookingSubmit,
    richTitle,
    authorDisplayName,
    handleBookButtonClick,
    handleMobileBookModalClose,
    onManageDisableScrolling,
  } = props;
  const showClosedListingHelpText = listing.id && isClosed;
  return (
    <div>
      <ModalInMobile
        className={css.modalInMobile}
        containerClassName={css.modalContainer}
        id="BookingDatesFormInModal"
        isModalOpenOnMobile={isBook}
        onClose={handleMobileBookModalClose}
        showAsModalMaxWidth={MODAL_BREAKPOINT}
        onManageDisableScrolling={onManageDisableScrolling}
      >
        <div className={css.modalHeading}>
          <h1 className={css.title}>{richTitle}</h1>
          <div className={css.author}>
            <span className={css.authorName}>
              <FormattedMessage id="ListingPage.hostedBy" values={{ name: authorDisplayName }} />
            </span>
          </div>
        </div>

        <div className={css.bookingHeading}>
          <h2 className={css.bookingTitle}>
            <FormattedMessage id="ListingPage.bookingTitle" values={{ title: richTitle }} />
          </h2>
          <div className={css.bookingHelp}>
            <FormattedMessage
              id={
                showClosedListingHelpText
                  ? 'ListingPage.bookingHelpClosedListing'
                  : 'ListingPage.bookingHelp'
              }
            />
          </div>
        </div>
        {!isClosed ? (
          <BookingDatesForm
            className={css.bookingForm}
            submitButtonWrapperClassName={css.bookingDatesSubmitButtonWrapper}
            unitType={unitType}
            onSubmit={handleBookingSubmit}
            price={price}
            isOwnListing={isOwnListing}
          />
        ) : null}
      </ModalInMobile>
      <div className={css.openBookingForm}>
        <div className={css.priceContainer}>
          <div className={css.priceValue} title={priceTitle}>
            {formattedPrice}
          </div>
          <div className={css.perUnit}>
            <FormattedMessage id="ListingPage.perUnit" />
          </div>
        </div>

        {!isClosed ? (
          <Button rootClassName={css.bookButton} onClick={handleBookButtonClick}>
            <FormattedMessage id="ListingPage.ctaButtonMessage" />
          </Button>
        ) : (
          <div className={css.closedListingButton}>
            <FormattedMessage id="ListingPage.closedListingButtonText" />
          </div>
        )}
      </div>
    </div>
  );
};

export default SectionBooking;
