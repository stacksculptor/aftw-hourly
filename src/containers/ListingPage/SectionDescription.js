import React from 'react';
import { FormattedMessage } from 'react-intl';
import { richText } from '../../util/richText';

import css from './ListingPage.css';

const MIN_LENGTH_FOR_LONG_WORDS_IN_DESCRIPTION = 20;

const SectionDescription = props => {
  const { description } = props;
  return (
    <div className={css.sectionDescription}>
      <h2 className={css.descriptionTitle}>
        <FormattedMessage id="ListingPage.descriptionTitle" />
      </h2>
      <p className={css.description}>
        {richText(description, {
          longWordMinLength: MIN_LENGTH_FOR_LONG_WORDS_IN_DESCRIPTION,
          longWordClass: css.longWord,
        })}
      </p>
    </div>
  );
};

export default SectionDescription;
