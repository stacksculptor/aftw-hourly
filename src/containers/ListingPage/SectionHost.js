import React from 'react';
import { FormattedMessage } from 'react-intl';
import { NamedLink, UserCard, Modal } from '../../components';
import { EnquiryForm } from '../../forms';

import css from './ListingPage.css';

const SectionHost = props => {
  const {
    title,
    listing,
    isOwnListing,
    authorDisplayName,
    onContactUser,
    isEnquiryModalOpen,
    onCloseEnquiryModal,
    sendEnquiryError,
    sendEnquiryInProgress,
    onSubmitEnquiry,
    currentUser,
    onManageDisableScrolling,
  } = props;
  return (
    <div id="host" className={css.sectionHost}>
      <h2 className={css.yourHostHeading}>
        <FormattedMessage id="ListingPage.yourHostHeading" />
      </h2>
      {isOwnListing ? (
        <NamedLink className={css.editProfileLink} name="ProfileSettingsPage">
          <FormattedMessage id="ListingPage.editProfileLink" />
        </NamedLink>
      ) : null}
      <UserCard user={listing.author} currentUser={currentUser} onContactUser={onContactUser} />
      <Modal
        id="ListingPage.enquiry"
        contentClassName={css.enquiryModalContent}
        isOpen={isEnquiryModalOpen}
        onClose={onCloseEnquiryModal}
        onManageDisableScrolling={onManageDisableScrolling}
      >
        <EnquiryForm
          className={css.enquiryForm}
          submitButtonWrapperClassName={css.enquirySubmitButtonWrapper}
          listingTitle={title}
          authorDisplayName={authorDisplayName}
          sendEnquiryError={sendEnquiryError}
          onSubmit={onSubmitEnquiry}
          inProgress={sendEnquiryInProgress}
        />
      </Modal>
    </div>
  );
};

export default SectionHost;
