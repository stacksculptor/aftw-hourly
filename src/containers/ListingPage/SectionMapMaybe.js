import React, { Component } from 'react';
import { string } from 'prop-types';
import { FormattedMessage } from 'react-intl';
import classNames from 'classnames';
import { propTypes } from '../../util/types';
import { obfuscatedCoordinates } from '../../util/maps';
import { Map } from '../../components';
import config from '../../config';

import css from './ListingPage.css';

class SectionMapMaybe extends Component {
  constructor(props) {
    super(props);
    this.state = { isStatic: true };
  }

  render() {
    const { className, rootClassName, geolocation, publicData, listingId } = this.props;

    if (!geolocation) {
      return null;
    }

    const address = publicData.location ? publicData.location.address : '';
    const classes = classNames(rootClassName || css.sectionMap, className);

    const mapProps = config.coordinates.fuzzy
      ? { obfuscatedCenter: obfuscatedCoordinates(geolocation, listingId ? listingId.uuid : null) }
      : { address, center: geolocation };

    return (
      <div className={classes}>
        <h2 className={css.locationTitle}>
          <FormattedMessage id="ListingPage.locationTitle" />
        </h2>
        <button
          className={css.map}
          onClick={() => {
            this.setState({ isStatic: false });
          }}
        >
          <Map {...mapProps} useStaticMap={this.state.isStatic} />
        </button>
      </div>
    );
  }
}

SectionMapMaybe.defaultProps = {
  rootClassName: null,
  className: null,
  geolocation: null,
  listingId: null,
};

SectionMapMaybe.propTypes = {
  rootClassName: string,
  className: string,
  geolocation: propTypes.latlng,
  listingId: propTypes.uuid,
};

export default SectionMapMaybe;
