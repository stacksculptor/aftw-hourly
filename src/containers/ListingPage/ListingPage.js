import React, { Component, PropTypes } from 'react';
import { FormattedMessage, intlShape, injectIntl } from 'react-intl';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import classNames from 'classnames';
import config from '../../config';
import * as propTypes from '../../util/propTypes';
import { types } from '../../util/sdkLoader';
import { createSlug } from '../../util/urlHelpers';
import { convertMoneyToNumber } from '../../util/currency';
import { createResourceLocatorString, findRouteByRouteName } from '../../util/routes';
import { ensureListing, ensureUser } from '../../util/data';
import { Avatar, Button, Map, ModalInMobile, PageLayout, ResponsiveImage } from '../../components';
import { BookingDatesForm, Topbar } from '../../containers';
import { getListingsById } from '../../ducks/marketplaceData.duck';
import { manageDisableScrolling, isScrollingDisabled } from '../../ducks/UI.duck';
import { showListing } from './ListingPage.duck';

import css from './ListingPage.css';

// This defines when ModalInMobile shows content as Modal
const MODAL_BREAKPOINT = 2500;

const { UUID } = types;

const priceData = (price, currencyConfig, intl) => {
  if (price && price.currency === currencyConfig.currency) {
    const priceAsNumber = convertMoneyToNumber(price, currencyConfig.subUnitDivisor);
    const formattedPrice = intl.formatNumber(priceAsNumber, currencyConfig);
    return { formattedPrice, priceTitle: formattedPrice };
  } else if (price) {
    return {
      formattedPrice: `(${price.currency})`,
      priceTitle: `Unsupported currency (${price.currency})`,
    };
  }
  return {};
};

// TODO: price unit (per x), custom fields, contact, reviews
// N.B. All the presentational content needs to be extracted to their own components
export class ListingPageComponent extends Component {
  constructor(props) {
    super(props);
    const tab = props.tab;
    this.state = {
      isBookingModalOpenOnMobile: tab && tab === 'book',
      pageClassNames: [],
    };

    this.onSubmit = this.onSubmit.bind(this);
  }

  onSubmit(values) {
    const { flattenedRoutes, history, getListing, params, useInitialValues } = this.props;
    const listing = getListing(new UUID(params.id));

    this.setState({ isBookingModalOpenOnMobile: false });

    // Customize checkout page state with current listing and selected bookingDates
    const { setInitialValues } = findRouteByRouteName('CheckoutPage', flattenedRoutes);
    const { startDate: bookingStart, endDate: bookingEnd } = values.bookingDates;
    const bookingDates = { bookingStart, bookingEnd };
    useInitialValues(setInitialValues, { listing, bookingDates, initiateOrderError: null });

    // Redirect to CheckoutPage
    history.push(
      createResourceLocatorString(
        'CheckoutPage',
        flattenedRoutes,
        { id: listing.id.uuid, slug: createSlug(listing.attributes.title) },
        {}
      )
    );
  }

  render() {
    const {
      history,
      location,
      params,
      showListingError,
      intl,
      currentUser,
      getListing,
      onManageDisableScrolling,
      scrollingDisabled,
    } = this.props;
    const currencyConfig = config.currencyConfig;
    const currentListing = ensureListing(getListing(new UUID(params.id)));
    const {
      address = '',
      description = '',
      geolocation = null,
      price = null,
      title = '',
    } = currentListing.attributes;

    if (!currentListing.id && showListingError) {
      const noDataMsg = { id: 'ListingPage.noListingData' };
      return <PageLayout title={intl.formatMessage(noDataMsg)} />;
    } else if (!currentListing.id) {
      const loadingPageMsg = { id: 'ListingPage.loadingListingData' };
      return <PageLayout title={intl.formatMessage(loadingPageMsg)} />;
    }

    const firstImage = currentListing.images && currentListing.images.length > 0
      ? currentListing.images[0]
      : null;

    const authorAvailable = currentListing && currentListing.author;
    const userAndListingAuthorAvailable = currentUser && authorAvailable;
    const isOwnListing = userAndListingAuthorAvailable &&
      currentListing.author.id.uuid === currentUser.id.uuid;
    const showBookButton = !isOwnListing;

    const { firstName: authorFirstName, lastName: authorLastName } = authorAvailable
      ? ensureUser(currentListing.author).attributes.profile
      : { firstName: '', lastName: '' };

    // TODO location address is currently serialized inside address field (API will change later)
    // Content is something like { locationAddress: 'Street, Province, Country', building: 'A 42' };
    let locationAddress = '';
    try {
      const deserializedAddress = JSON.parse(address || '{}');
      locationAddress = deserializedAddress.locationAddress;
    } catch (e) {
      locationAddress = address;
    }

    const bookBtnMessage = intl.formatMessage({ id: 'ListingPage.ctaButtonMessage' });
    const { formattedPrice, priceTitle } = priceData(price, currencyConfig, intl);
    const map = geolocation
      ? <div className={css.locationContainer}>
          <h3 className={css.locationTitle}>
            <FormattedMessage id="ListingPage.locationTitle" />
          </h3>
          <div className={css.map}><Map center={geolocation} address={locationAddress} /></div>
        </div>
      : null;

    const listingClasses = classNames(css.listing, { [css.bookable]: showBookButton });

    return (
      <PageLayout title={`${title} ${formattedPrice}`} scrollingDisabled={scrollingDisabled}>
        <Topbar history={history} location={location} />
        <div className={listingClasses}>
          <div className={css.threeToTwoWrapper}>
            <div className={css.aspectWrapper}>
              <ResponsiveImage
                rootClassName={css.rootForImage}
                alt={title}
                image={firstImage}
                nameSet={[
                  { name: 'landscape-crop', size: '400w' },
                  { name: 'landscape-crop2x', size: '800w' },
                ]}
                sizes="100vw"
              />
            </div>
          </div>

          <div className={css.avatarWrapper}>
            <Avatar
              rootClassName={css.avatar}
              firstName={authorFirstName}
              lastName={authorLastName}
            />
          </div>

          <div className={css.heading}>
            <h1 className={css.title}>{title}</h1>
            <div className={css.author}>
              <span className={css.authorName}>
                <FormattedMessage id="ListingPage.hostedBy" values={{ name: authorFirstName }} />
              </span>
            </div>
          </div>

          <div className={css.descriptionContainer}>
            <h3 className={css.descriptionTitle}>
              <FormattedMessage id="ListingPage.descriptionTitle" />
            </h3>
            <p className={css.description}>{description}</p>
          </div>

          {map}

          <ModalInMobile
            id="BookingDatesFormInModal"
            isModalOpenOnMobile={this.state.isBookingModalOpenOnMobile}
            onClose={() => this.setState({ isBookingModalOpenOnMobile: false })}
            showAsModalMaxWidth={MODAL_BREAKPOINT}
            onManageDisableScrolling={onManageDisableScrolling}
          >
            <div className={css.modalHeading}>
              <h1 className={css.title}>{title}</h1>
              <div className={css.author}>
                <span className={css.authorName}>
                  <FormattedMessage id="ListingPage.hostedBy" values={{ name: authorFirstName }} />
                </span>
              </div>
            </div>
            <BookingDatesForm className={css.bookingForm} onSubmit={this.onSubmit} price={price} />
          </ModalInMobile>
          {showBookButton
            ? <div className={css.openBookingForm}>
                <div className={css.priceContainer}>
                  <div className={css.priceValue} title={priceTitle}>
                    {formattedPrice}
                  </div>
                  <div className={css.perNight}>
                    <FormattedMessage id="ListingPage.perNight" />
                  </div>
                </div>

                <Button
                  rootClassName={css.bookButton}
                  onClick={() => this.setState({ isBookingModalOpenOnMobile: true })}
                >
                  {bookBtnMessage}
                </Button>
              </div>
            : null}
        </div>
      </PageLayout>
    );
  }
}

ListingPageComponent.defaultProps = {
  showListingError: null,
  tab: 'listing',
  currentUser: null,
};

const { arrayOf, bool, func, instanceOf, object, oneOf, shape, string } = PropTypes;

ListingPageComponent.propTypes = {
  // from withRouter
  history: shape({
    push: func.isRequired,
  }).isRequired,
  location: object.isRequired,
  flattenedRoutes: arrayOf(propTypes.route).isRequired,
  // from injectIntl
  intl: intlShape.isRequired,
  params: shape({
    id: string.isRequired,
    slug: string.isRequired,
  }).isRequired,
  showListingError: instanceOf(Error),
  currentUser: propTypes.currentUser,
  getListing: func.isRequired,
  tab: oneOf(['book', 'listing']),
  scrollingDisabled: bool.isRequired,
  onManageDisableScrolling: func.isRequired,
  useInitialValues: func.isRequired,
};

const mapStateToProps = state => {
  const { showListingError } = state.ListingPage;
  const { currentUser } = state.user;

  const getListing = id => {
    const listings = getListingsById(state, [id]);
    return listings.length === 1 ? listings[0] : null;
  };

  return {
    showListingError,
    currentUser,
    getListing,
    scrollingDisabled: isScrollingDisabled(state),
  };
};

const mapDispatchToProps = dispatch => ({
  onManageDisableScrolling: (componentId, disableScrolling) =>
    dispatch(manageDisableScrolling(componentId, disableScrolling)),
  useInitialValues: (setInitialValues, { listing, bookingDates, initiateOrderError }) =>
    dispatch(setInitialValues({ listing, bookingDates, initiateOrderError })),
});

const ListingPage = compose(connect(mapStateToProps, mapDispatchToProps), withRouter, injectIntl)(
  ListingPageComponent
);

ListingPage.loadData = params => {
  const id = new UUID(params.id);
  return showListing(id);
};

export default ListingPage;
