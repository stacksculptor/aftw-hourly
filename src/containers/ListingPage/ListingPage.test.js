import React from 'react';
import { shallow } from 'enzyme';
import { FormattedMessage } from 'react-intl';
import { types } from '../../util/sdkLoader';
import { createUser, createCurrentUser, createListing, fakeIntl } from '../../util/test-data';
import { renderShallow } from '../../util/test-helpers';
import { ListingPageComponent, ActionBar } from './ListingPage';
import { addMarketplaceEntities } from '../../ducks/marketplaceData.duck';
import { showListingRequest, showListingError, showListing } from './ListingPage.duck';

const { UUID } = types;
const noop = () => null;

describe('ListingPage', () => {
  it('matches snapshot', () => {
    const currentUser = createCurrentUser('user-2');
    const listing1 = createListing('listing1', {}, { author: createUser('user-1') });
    const getListing = () => listing1;
    const props = {
      flattenedRoutes: [],
      location: { search: '' },
      history: {
        push: () => console.log('HistoryPush called'),
      },
      params: { slug: 'listing1-title', id: 'listing1' },
      currentUser,
      getListing: getListing,
      intl: fakeIntl,
      authInProgress: false,
      currentUserHasListings: false,
      isAuthenticated: false,
      onLogout: noop,
      onLoadListing: noop,
      onManageDisableScrolling: noop,
      scrollingDisabled: false,
      useInitialValues: noop,
      sendVerificationEmailInProgress: false,
      onResendVerificationEmail: noop,
    };

    const tree = renderShallow(<ListingPageComponent {...props} />);
    expect(tree).toMatchSnapshot();
  });

  describe('Duck', () => {
    it('showListing() success', () => {
      const id = new UUID('00000000-0000-0000-0000-000000000000');
      const dispatch = jest.fn(action => action);
      const response = { status: 200 };
      const show = jest.fn(() => Promise.resolve(response));
      const sdk = { listings: { show }, currentUser: { show } };

      return showListing(id)(dispatch, null, sdk).then(data => {
        expect(data).toEqual(response);
        expect(show.mock.calls).toEqual([[{ id, include: ['author', 'images'] }]]);
        expect(dispatch.mock.calls).toEqual([
          [showListingRequest(id)],
          [expect.anything()], // fetchCurrentUser() call
          [addMarketplaceEntities(data)],
        ]);
      });
    });

    it('showListing() error', () => {
      const id = new UUID('00000000-0000-0000-0000-000000000000');
      const dispatch = jest.fn(action => action);
      const error = new Error('fail');
      const show = jest.fn(() => Promise.reject(error));
      const sdk = { listings: { show } };

      // Calling sdk.listings.show is expected to fail now

      return showListing(id)(dispatch, null, sdk).then(
        () => {
          throw new Error('sdk.listings.show was supposed to fail!');
        },
        e => {
          expect(e).toEqual(error);
          expect(show.mock.calls).toEqual([[{ id, include: ['author', 'images'] }]]);
          expect(dispatch.mock.calls).toEqual([
            [showListingRequest(id)],
            [expect.anything()], // fetchCurrentUser() call
            [showListingError(e)],
          ]);
        }
      );
    });
  });

  describe('ActionBar', () => {
    it('shows users own listing status', () => {
      const actionBar = shallow(<ActionBar isOwnListing isClosed={false} editParams={{}} />);
      const formattedMessages = actionBar.find(FormattedMessage);
      expect(formattedMessages.length).toEqual(2);
      expect(formattedMessages.at(0).props().id).toEqual('ListingPage.ownListing');
      expect(formattedMessages.at(1).props().id).toEqual('ListingPage.editListing');
    });
    it('shows users own closed listing status', () => {
      const actionBar = shallow(<ActionBar isOwnListing isClosed editParams={{}} />);
      const formattedMessages = actionBar.find(FormattedMessage);
      expect(formattedMessages.length).toEqual(2);
      expect(formattedMessages.at(0).props().id).toEqual('ListingPage.ownClosedListing');
      expect(formattedMessages.at(1).props().id).toEqual('ListingPage.editListing');
    });
    it('shows closed listing status', () => {
      const actionBar = shallow(<ActionBar isOwnListing={false} isClosed editParams={{}} />);
      const formattedMessages = actionBar.find(FormattedMessage);
      expect(formattedMessages.length).toEqual(1);
      expect(formattedMessages.at(0).props().id).toEqual('ListingPage.closedListing');
    });
    it("is missing if listing is not closed or user's own", () => {
      const actionBar = shallow(
        <ActionBar isOwnListing={false} isClosed={false} editParams={{}} />
      );
      expect(actionBar.equals(null)).toEqual(true);
    });
  });
});
