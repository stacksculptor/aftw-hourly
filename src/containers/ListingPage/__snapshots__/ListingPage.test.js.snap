// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingPage matches snapshot 1`] = `
<Page
  authInfoError={null}
  author="user-1 display name"
  canonicalPath="/l/listing1"
  contentType="website"
  description="listing1 description"
  facebookImages={Array []}
  logoutError={null}
  schema="
          {
            \\"@context\\": \\"http://schema.org\\",
            \\"@type\\": \\"ItemPage\\",
            \\"description\\": \\"listing1 description\\",
            \\"name\\": \\"ListingPage.schemaTitle\\",
            \\"image\\": []
          }
        "
  scrollingDisabled={false}
  title="ListingPage.schemaTitle"
  twitterImages={Array []}
>
  <Connect(withRouter(TopbarContainerComponent)) />
  <div
    className=""
  >
    <div>
      <div
        onClick={[Function]}
      >
        <div
          onClick={[Function]}
        >
          <ActionBar
            editParams={
              Object {
                "id": "listing1",
                "slug": "listing1-title",
                "tab": "description",
                "type": "edit",
              }
            }
            isClosed={false}
            isOwnListing={false}
          />
        </div>
        <ResponsiveImage
          alt="listing1 title"
          className={null}
          image={null}
          nameSet={
            Array [
              Object {
                "name": "landscape-crop",
                "size": "400w",
              },
              Object {
                "name": "landscape-crop2x",
                "size": "800w",
              },
              Object {
                "name": "landscape-crop4x",
                "size": "1600w",
              },
              Object {
                "name": "landscape-crop6x",
                "size": "2400w",
              },
            ]
          }
          noImageMessage={null}
          rootClassName={null}
          sizes="100vw"
        />
      </div>
    </div>
    <InjectIntl(ModalComponent)
      id="ListingPage.imageCarousel"
      isOpen={false}
      lightCloseButton={true}
      onClose={[Function]}
      onManageDisableScrolling={[Function]}
    >
      <InjectIntl(ImageCarousel)
        images={Array []}
      />
    </InjectIntl(ModalComponent)>
    <div>
      <div>
        <AvatarLarge
          user={
            Object {
              "attributes": Object {
                "banned": false,
                "profile": Object {
                  "abbreviatedName": "user-1 abbreviated name",
                  "displayName": "user-1 display name",
                },
              },
              "id": UUID {
                "uuid": "user-1",
              },
              "type": "user",
            }
          }
        />
        <AvatarMedium
          user={
            Object {
              "attributes": Object {
                "banned": false,
                "profile": Object {
                  "abbreviatedName": "user-1 abbreviated name",
                  "displayName": "user-1 display name",
                },
              },
              "id": UUID {
                "uuid": "user-1",
              },
              "type": "user",
            }
          }
        />
      </div>
      <div>
        <div>
          <div>
            <div
              title={55}
            >
              55
            </div>
            <div>
              <FormattedMessage
                id="ListingPage.perNight"
                values={Object {}}
              />
            </div>
          </div>
          <div>
            <h1>
              listing1 title
            </h1>
            <div>
              <span>
                <FormattedMessage
                  id="ListingPage.hostedBy"
                  values={
                    Object {
                      "name": "user-1 display name",
                    }
                  }
                />
              </span>
            </div>
          </div>
        </div>
        <div>
          <h2>
            <FormattedMessage
              id="ListingPage.descriptionTitle"
              values={Object {}}
            />
          </h2>
          <p>
            listing1 description
          </p>
        </div>
        <div>
          <h2>
            <FormattedMessage
              id="ListingPage.locationTitle"
              values={Object {}}
            />
          </h2>
          <div>
            <Map
              address="listing1 address"
              center={
                LatLng {
                  "lat": 40,
                  "lng": 60,
                }
              }
              className=""
              rootClassName={null}
              zoom={11}
            />
          </div>
        </div>
      </div>
      <withViewport(ModalInMobileComponent)
        id="BookingDatesFormInModal"
        isModalOpenOnMobile={false}
        onClose={[Function]}
        onManageDisableScrolling={[Function]}
        showAsModalMaxWidth={1023}
      >
        <div>
          <h1>
            listing1 title
          </h1>
          <div>
            <span>
              <FormattedMessage
                id="ListingPage.hostedBy"
                values={
                  Object {
                    "name": "user-1 display name",
                  }
                }
              />
            </span>
          </div>
        </div>
        <div>
          <h2>
            <FormattedMessage
              id="ListingPage.bookingTitle"
              values={
                Object {
                  "title": "listing1 title",
                }
              }
            />
          </h2>
          <div>
            <FormattedMessage
              id="ListingPage.bookingHelp"
              values={Object {}}
            />
          </div>
        </div>
        <BookingDatesForm
          isOwnListing={false}
          onSubmit={[Function]}
          price={
            Money {
              "amount": 5500,
              "currency": "USD",
            }
          }
        />
      </withViewport(ModalInMobileComponent)>
      <div>
        <div>
          <div
            title={55}
          >
            55
          </div>
          <div>
            <FormattedMessage
              id="ListingPage.perNight"
              values={Object {}}
            />
          </div>
        </div>
        <Button
          className={null}
          disabled={false}
          inProgress={false}
          onClick={[Function]}
          ready={false}
          rootClassName={null}
        >
          ListingPage.ctaButtonMessage
        </Button>
      </div>
    </div>
  </div>
</Page>
`;
